### PR TITLE
Gitlab factory authentication using  OAuth 2 flow

### PR DESCRIPTION
**Pull request by mshaposhnik**
### What does this PR do?
Enables Gitlab factory authentication using OAuth 2 flow

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/19171

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [ ] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [ ] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [ ] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [ ] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [ ] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

File: wsmaster/che-core-api-factory-gitlab/src/main/java/org/eclipse/che/api/factory/server/gitlab/GitlabOAuthTokenFetcher.java [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+import org.eclipse.che.api.factory.server.scm.exception.ScmItemNotFoundException;
+import org.eclipse.che.api.factory.server.scm.exception.ScmUnauthorizedException;
+import org.eclipse.che.commons.annotation.Nullable;
+import org.eclipse.che.commons.lang.NameGenerator;
+import org.eclipse.che.commons.subject.Subject;
+import org.eclipse.che.security.oauth.OAuthAPI;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** GitLab OAuth token retriever. */
```
Comment (skabashnyuk) [inline wsmaster/che-core-api-factory-gitlab/src/main/java/org/eclipse/che/api/factory/server/gitlab/GitlabOAuthTokenFetcher.java:41]: Test

**Review by skabashnyuk — COMMENTED** [review]
In general looks good.

File: wsmaster/che-core-api-factory-gitlab/src/main/java/org/eclipse/che/api/factory/server/gitlab/GitlabOAuthTokenFetcher.java [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+  private static final Logger LOG = LoggerFactory.getLogger(GitlabOAuthTokenFetcher.class);
+  private static final String OAUTH_PROVIDER_NAME = "gitlab";
+
+  private final OAuthAPI oAuthAPI;
+  private final String apiEndpoint;
+  private final GitlabApiClient gitlabApiClient;
+
+  @Inject
+  public GitlabOAuthTokenFetcher(
+      @Nullable @Named("che.integration.gitlab.server_endpoints") String gitlabEndpoints,
```
Comment (skabashnyuk) [inline wsmaster/che-core-api-factory-gitlab/src/main/java/org/eclipse/che/api/factory/server/gitlab/GitlabOAuthTokenFetcher.java:55]: che.integration.gitlab.server_endpoints - can't see defaults and description.

File: wsmaster/che-core-api-factory-gitlab/src/main/java/org/eclipse/che/api/factory/server/gitlab/GitlabApiClient.java [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+/** GitLab API operations helper. */
+public class GitlabApiClient {
+
+  private static final Logger LOG = LoggerFactory.getLogger(GitlabApiClient.class);
+
+  private final HttpClient httpClient;
+  private final URI serverUrl;
+
+  private static final Duration DEFAULT_HTTP_TIMEOUT = ofSeconds(10);
+  private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
```
Comment (metlos) [inline wsmaster/che-core-api-factory-gitlab/src/main/java/org/eclipse/che/api/factory/server/gitlab/GitlabApiClient.java:49]: Are we sure object mapper is thread safe?

Comment (mshaposhnik) [reply wsmaster/che-core-api-factory-gitlab/src/main/java/org/eclipse/che/api/factory/server/gitlab/GitlabApiClient.java:49]: Looks like [yes](https://stackoverflow.com/questions/3907929/should-i-declare-jacksons-objectmapper-as-a-static-field)

File: wsmaster/che-core-api-factory-gitlab/src/main/java/org/eclipse/che/api/factory/server/gitlab/GitlabApiClient.java [hunk 4]
```diff
@@ -84,6 +84,31 @@ public GitlabUser getUser(String authenticationToken)
         });
   }
 
+  public GitlabOauthTokenInfo tokenInfo(String authenticationToken)
```
Comment (skabashnyuk) [inline wsmaster/che-core-api-factory-gitlab/src/main/java/org/eclipse/che/api/factory/server/gitlab/GitlabApiClient.java]: tokenInfo -> getTokenInfo?

File: wsmaster/che-core-api-factory-gitlab/src/main/java/org/eclipse/che/api/factory/server/gitlab/GitlabOAuthTokenFetcher.java [hunk 5]
```diff
@@ -103,6 +105,24 @@ public PersonalAccessToken fetchPersonalAccessToken(Subject cheSubject, String s
...
+    if (gitlabApiClient == null
+        || !gitlabApiClient.isConnected(personalAccessToken.getScmProviderUrl())) {
+      LOG.debug(
+          "not a  valid url {} for current fetcher ", personalAccessToken.getScmProviderUrl());
+      return Optional.empty();
+    }
+    try {
+      GitlabOauthTokenInfo info = gitlabApiClient.tokenInfo(personalAccessToken.getToken());
+      return Optional.of(
+          Sets.difference(DEFAULT_TOKEN_SCOPE, Sets.newHashSet(info.getScope())).isEmpty());
```
Comment (skabashnyuk) [inline wsmaster/che-core-api-factory-gitlab/src/main/java/org/eclipse/che/api/factory/server/gitlab/GitlabOAuthTokenFetcher.java]: set equals ? same as in BitbucketServerPersonalAccessTokenFetcher .

Comment (mshaposhnik) [reply wsmaster/che-core-api-factory-gitlab/src/main/java/org/eclipse/che/api/factory/server/gitlab/GitlabOAuthTokenFetcher.java]: imo it's wrong. scopes must be rule `not-less-than` of what we require.

**Review by skabashnyuk — APPROVED** [review]

**Pull request merged**